torch/csrc/jit/frontend/ir_emitter_utils.h

```diff
@@ -0,0 +1,421 @@
+
+void checkApplyNumInputs(Apply& apply, size_t expected_inputs) {
+  const SourceRange& loc = apply.range();
+  if (apply.inputs().size() != expected_inputs) {
+    throw ErrorReport(loc) << Var(apply.callee()).name().name()
+                           << " expected exactly " << expected_inputs
+                           << " arguments but found " << apply.inputs().size();
+  }
+  if (apply.attributes().size() > 0) {
+    throw ErrorReport(loc) << Var(apply.callee()).name().name()
+                           << " takes no keyword arguments";
+  }
+}
+
+inline bool isSupportedListElementType(const TypePtr& type) {
+  return type->isSubtypeOf(TensorType::get()) ||
+      type->isSubtypeOf(NumberType::get());
+}
+
+// Validate that the `lhs` Expr's in an assignment statement are valid. That
+// is:
+//
+// 1) All lhs Expr's are either Var, Tuple or Starred nodes
+// 2) There is at most one Starred node in the lhs Expr
+// 3) A Starred node can only appear when there is another non-Starred lhs
+//    Expr. Concretely this means that `*abc = func()` is illegal. Unpacking
+//    all outputs into a tuple is covered by `abc = func()`.
+bool validateAssignLhsExpr(const List<Expr>& lhs, const SourceRange& r) {
+  size_t num_normal_assign = 0;
+  size_t num_starred = 0;
+  for (const auto& assignee : lhs) {
+    if (assignee.kind() == TK_VAR || assignee.kind() == TK_SUBSCRIPT ||
+        assignee.kind() == TK_TUPLE_LITERAL || assignee.kind() == '.') {
+      num_normal_assign++;
+    } else if (assignee.kind() == TK_STARRED) {
+      num_starred++;
+    } else {
+      throw ErrorReport(assignee) << "lhs of assignment must be a variable, "
+                                  << "subscript, or starred expression";
+    }
+  }
+
+  if (num_starred > 1) {
+    throw ErrorReport(r) << "Only one starred expression is allowed on the lhs";
+  }
+
+  if (num_starred > 0 && num_normal_assign == 0) {
+    throw ErrorReport(r) << "A Starred expression may only appear on the "
+                         << "lhs within the presence of another non-starred"
+                         << " expression";
+  }
+
+  return num_starred;
+}
+
+/* ============================================ */
+/*                      Casts                   */
+/* ============================================ */
+
+static Value* asSimple(const SugaredValuePtr& value) {
+  if (SimpleValue* sv = dynamic_cast<SimpleValue*>(value.get())) {
+    return sv->getValue();
+  }
+  return nullptr;
+}
+
+static std::shared_ptr<MagicMethod> makeMagic(
+    const std::string& name,
+    SugaredValuePtr base) {
+  return std::make_shared<MagicMethod>(name, base);
+}
+
+/* ==================================================================== */
+/*                      `__setstate__` Information                      */
+/* ==================================================================== */
+
+// see [setstate type]
+static TypePtr getTypeForSetStateArg(const Def& def, const Self* self) {
```

Comment:
`static` will cause function definitions to be copied N times if this header is included in N different translation units (files). If we meant to share this function through the header, then consider removing `static`, otherwise if this is logically private we should move this in an implementation file.

Reply:
I think we can remove `static` altogether. It was used in `ir_emitter.cpp` because this function was defined within the scope of `to_ir` but was not a member function.